docs/proposals/20200506-conditions.md

```diff
@@ -0,0 +1,363 @@
+  control plane, and other related objects.
+
+The final iteration of this proposal will focus on:
+
+1. To define a mechanism for bubbling up/aggregating conditions up to hierarchy of cluster
+   objects
+
+
+### Non-Goals/Future Work
+
+Non-Goals
+
+- To pre-define every possible condition.
+- To remove  `Phase`, `ErrorMessage`, `ErrorReason` from Status in v0.3 .
+- To modify any Cluster API code in v0.3 to rely on conditions to determine actions or workflows.
+
+Future Work (v1alpha4)
+
+- To add core conditions in the Cluster API contract, eventually deprecating `Phase`, `ErrorMessage`,
+  `ErrorReason` from Status (TBD).
+- To modify Cluster API code to rely on conditions to determine actions or workflows.
+
+## Proposal
+
+Disclaimer: the goal of this first iteration is to to agree on a design for the condition types. 
+The condition types included in this document are only meant to validate the type design. There are
+no guarantees such examples will become real conditions during next iterations.
+
+### User Stories
+
+- Detail the things that people will be able to do if this proposal is implemented.
+- Include as much detail as possible so that people can understand the "how" of the system.
+- The goal here is to make this feel real for users without getting bogged down.
+
+#### Story 1
+
+As a developer I would like to have common types for defining conditions on different Cluster API
+or provider objects.
+
+### Implementation Details/Notes/Constraints
+
+This proposal aims to be consistent with the target state of conditions in Kubernetes (see [KEP](https://github.com/kubernetes/enhancements/pull/1624)).
+
+At the same time, we should think that Cluster API presents some specific challenges that are not common
+to the core Kubernetes objects:
+
+- Cluster API has a high number of long-running operations, and this makes it crucial for the success of
+  this proposal to introduce a clean way to provide evidence about what is happening during those operations,
+  and not only to show the final state of the operations.
+- Cluster API has a complex hierarchy of objects, and the informative value of a condition, especially in
+  the core types, is highly correlated to the possibility to collect and summarize meaningful information
+  from the underlying hierarchy of objects.
+- Cluster API defines a set of core objects and extension points (e.g. bootstrap, control plane and infrastructure).
+  Extension points are defined by a set of contracts, which each provider adheres to. To provide a great user
+  experience, core and provider contracts should be extended to allow for better coordination and visibility
+  based on conditions.
+
+#### Data Model Changes
+
+In this section we outline the proposed API changes.
+
+All the changes can be implemented in the v0.3.x timeframe (no breaking changes, only additions).
+
+Following types should be defined in Cluster API / v1alpha3 api.
+
```

Comment:
I think that to figure out a data model and possibly justify skew from https://github.com/kubernetes/enhancements/pull/1624 we need to first flesh out some user stories traversing the different CAPI components to understand which data model can best convey our contents.

Reply:
Done, user stories and examples are now included